In [1]:
import pprint
import sqlite3
import pandas as pd
from IPython.display import display

import os
from hurry.filesize import size


conn = sqlite3.connect('swlondon.sqlite')
cur = conn.cursor()

In [2]:
# ONLY RUN ONCE PER SESSION!!

dirpath = '.'

file_size_wanted = ['nodes.csv','nodes_tags.csv','swlondon.osm','swlondon.sqlite',
                    'ways.csv','ways_nodes.csv','ways_tags.csv']

files_list = []
for path, dirs, files in os.walk(dirpath):
    files_list.extend([(filename, size(os.path.getsize(os.path.join(path, filename)))) for filename in files])

    

In [3]:
files_list

[('.DS_Store', '6K'),
 ('CBRAKE_osm_project_submission_2_280717.zip', '1M'),
 ('csv_to_database.py', '4K'),
 ('data_wrangling_schema.sql', '833B'),
 ('element_type_count.py', '643B'),
 ('Icon\r', '0B'),
 ('initial_street_name_audit.py', '1K'),
 ('k_attrib_audit.py', '626B'),
 ('map_area_selected.txt', '382B'),
 ('nodes.csv', '91M'),
 ('nodes_tags.csv', '6M'),
 ('osm_db_queries.ipynb', '97K'),
 ('osm_export_sg.jpeg', '714K'),
 ('osm_processing_code.ipynb', '28K'),
 ('osm_project_code.ipynb', '244K'),
 ('osm_project_notes.ipynb', '38K'),
 ('OSM_project_submission.ipynb', '26K'),
 ('OSM_project_submission_2.html', '269K'),
 ('osm_sampling.py', '948B'),
 ('postcode_audit.py', '2K'),
 ('process_to_csv.py', '12K'),
 ('READ ME.txt', '1K'),
 ('README.md', '2K'),
 ('schema.py', '2K'),
 ('schema.pyc', '1K'),
 ('street_name_audit.py', '3K'),
 ('swlondon.osm', '264M'),
 ('swlondon.sqlite', '138M'),
 ('swlondon_sample_submission.osm', '10M'),
 ('tag_type_review.py', '1K'),
 ('ways.csv', '10M'),
 ('

In [4]:
files_list.sort(key=lambda size: int(size[1].translate(None, "MKB")), reverse=True)

files_list

[('9435cd8a4105ca7b57b59867a1ea4cad75623a', '979B'),
 ('8bd0ae81f9ff6e82e936488d665840250b2d55', '960B'),
 ('osm_sampling.py', '948B'),
 ('osm_sampling.py', '948B'),
 ('osm_sampling.py', '948B'),
 ('commit-msg.sample', '896B'),
 ('master', '856B'),
 ('data_wrangling_schema.sql', '833B'),
 ('schema.cpython-36.pyc', '810B'),
 ('39d08a5c591fc05f7a21be7c3ff19759cc5f5c', '807B'),
 ('920d47c7bfea55c7bc6340d11295c945012925', '740B'),
 ('osm_export_sg.jpeg', '714K'),
 ('c586671ff3474a1cc931ad8b99a1b0f9099117', '699B'),
 ('0a1eac7b1b2d3acab400d46de3000fb429b7b2', '668B'),
 ('element_type_count.py', '643B'),
 ('element_type_count.py', '643B'),
 ('element_type_count.py', '643B'),
 ('k_attrib_audit.py', '626B'),
 ('k_attrib_audit.py', '626B'),
 ('k_attrib_audit.py', '626B'),
 ('9338ac10793aced5c103c6d6c93e7c4ccbf936', '606B'),
 ('13807a76528f1c34a380c5a98b7ea7619ec246', '605B'),
 ('f2c2fc54589e650d90c7241d43f636fd6cc146', '578B'),
 ('22290c738d830e0b8ca3f275bd7e8fcaafd547', '575B'),
 ('2c6bb29a116

In [5]:
file_size_wanted = ['nodes.csv','nodes_tags.csv','swlondon.osm','swlondon.sqlite',
                    'ways.csv','ways_nodes.csv','ways_tags.csv']


for filename, size in files_list:
    if filename in file_size_wanted :
        print '{:.<40s}: {:5s}'.format(filename,size)

swlondon.osm............................: 264M 
swlondon.sqlite.........................: 138M 
nodes.csv...............................: 91M  
ways_nodes.csv..........................: 33M  
ways_tags.csv...........................: 16M  
ways.csv................................: 10M  
nodes_tags.csv..........................: 6M   


In [6]:
df = pd.read_sql_query(
''' 
SELECT * from nodes
LIMIT 3
''' , conn)


print ('nodes table')
display(df)
print  ('\n')


df2 = pd.read_sql_query(
''' 
SELECT * from ways
LIMIT 3
''' , conn)

print ('ways table')
display(df2)
print  ('\n')

df3 = pd.read_sql_query(
''' 
SELECT * from nodes_tags
LIMIT 3
''' , conn)

print ('nodes_tags table')
display(df3)
print  ('\n')


df4 = pd.read_sql_query(
''' 
SELECT * from ways_tags
LIMIT 3
''' , conn)

print ('ways_tags table')
display(df4)
print  ('\n')

df5 = pd.read_sql_query(
''' 
SELECT * from ways_nodes
LIMIT 3
''' , conn)

print ('ways_nodes table')
display(df5)
print  ('\n')

nodes table


id        lat       lon       user     uid  version  changeset  \
0  109642  51.483586 -0.167361  MacLondon  322039       10   43286497   
1  109644  51.482444 -0.172071  Firefishy    3560        5    1873462   
2  109645  51.482151 -0.173407  MacLondon  322039        5   29279984   

              timestamp  
0  2016-10-30T16:09:18Z  
1  2009-07-19T14:38:21Z  
2  2015-03-06T00:39:17Z



ways table


id        user     uid version  changeset             timestamp
0  2987  jpennycook   17292       8   49509164  2017-06-13T19:12:51Z
1  2989      alpman  364615       3   50655390  2017-07-28T18:00:39Z
2  2991    Wynndale  118262       3   12410104  2012-07-21T13:27:36Z



nodes_tags table


id             key                                              value  \
0  109644  randomjunk_bot  reverted edit by FilipeOliveira on 2008-11-24T...   
1  109645         highway                                    traffic_signals   
2  109645  randomjunk_bot  reverted edit by FilipeOliveira on 2008-11-24T...   

      type  
0  regular  
1  regular  
2  regular



ways_tags table


id      key        value     type
0  2987   access      private  regular
1  2987  highway      service  regular
2  2989  highway  residential  regular



ways_nodes table


id     node_id  position
0  2987  1667684820         0
1  2987  3927098721         1
2  2987  3927098722         2

In [7]:

# Count of Nodes & Ways

df1 = pd.read_sql_query(
''' 
SELECT COUNT(nodes.id) as Nodes
FROM nodes
''' , conn)

print ('Number of Nodes:')
display(df1)
print  ('\n')

df2 = pd.read_sql_query(
''' 
SELECT COUNT(ways.id) as Ways
FROM ways
''' , conn)

print ('Number of Ways:')
display(df2)
print  ('\n')



Number of Nodes:


Nodes
0  1174364



Number of Ways:


Ways
0  182804

In [8]:
# Count unique contributing users

df1 = pd.read_sql_query(
''' 
SELECT COUNT(DISTINCT(all_users.uid)) as UserCount          
FROM (SELECT uid FROM nodes UNION ALL SELECT uid FROM ways) all_users
''' , conn)

df1

UserCount
0       2165

In [9]:
# Count contributions from top 25 contributing users

df = pd.read_sql_query(
''' 
SELECT all_users.user, count(all_users.uid) as contributions
FROM (SELECT uid, user FROM nodes UNION ALL SELECT uid, user FROM ways) as all_users
GROUP BY all_users.uid
ORDER BY Contributions DESC
limit 15;
''' , conn)


df

user  contributions
0            Rondon237         183336
1              ecatmur         102746
2                  OJW          66467
3       joelholdsworth          59083
4          Ian Haylock          46787
5             UniEagle          33668
6               Mark_S          32565
7               moyogo          31530
8             Wynndale          28578
9              Ed Avis          26528
10           marsbar28          25082
11  Paul The Archivist          20271
12      Derick Rethans          19714
13   LivingWithDragons          19481
14        captainpovey          18194

In [10]:
# How many users have only made one contribution?

df = pd.read_sql_query(
''' 
SELECT COUNT(*) as one_contrib_users
FROM
    (SELECT all_users.uid, all_users.user, count(all_users.uid) as contributions
    FROM (SELECT uid, user FROM nodes UNION ALL SELECT uid, user FROM ways) as all_users
    GROUP BY all_users.uid
    HAVING contributions=1) as singles;

''' , conn)


df




one_contrib_users
0                473

In [11]:
df = pd.read_sql_query(
''' 
SELECT value, COUNT(*) as num
FROM nodes_tags
WHERE key='amenity'
GROUP BY value
ORDER BY num DESC
LIMIT 35;

''' , conn)


df

value   num
0           post_box  1614
1    bicycle_parking   801
2         restaurant   671
3              bench   655
4                pub   556
5               cafe   478
6          telephone   440
7          fast_food   332
8                atm   273
9   place_of_worship   265
10      waste_basket   226
11       car_sharing   159
12              bank   154
13          pharmacy   154
14           toilets   151
15         recycling   146
16           parking   130
17       post_office   125
18              fuel   108
19               bar    74
20           dentist    74
21    bicycle_rental    73
22   vending_machine    68
23            school    53
24           doctors    47
25           library    46
26      kindergarten    43
27          hospital    33
28    drinking_water    31
29          fountain    28
30              taxi    28
31  community_centre    23
32            police    22
33          grit_bin    20
34        veterinary    18

In [12]:
# List of Pub names


df = pd.read_sql_query(
''' 
SELECT pubs.id, pubs.value, names.value
FROM
    (SELECT id, value
    FROM nodes_tags
    WHERE key='amenity'
    AND value='pub') as pubs 
    JOIN 
    (SELECT id, value
    FROM nodes_tags
    WHERE key='name'
    AND type='regular') as names

ON pubs.id =  names.id
GROUP BY pubs.id

''' , conn)


df

pubs.id pubs.value            names.value
0      13353276        pub             The Harrow
1      17806375        pub              Green Man
2      18235090        pub           The Arab Boy
3      19828422        pub            The Anglers
4      20627020        pub      The Edmund Tylney
5      20692145        pub        The Penny Black
6      20695067        pub             The Plough
7      20973312        pub             Kings Arms
8      20973316        pub          The Waterside
9      21980919        pub          The Manor Inn
10     26108616        pub         Kingswood Arms
11     26422089        pub               The Mint
12     26422124        pub           The Woolpack
13     26461817        pub               The Goat
14     26462174        pub           The Red Lion
15     26545063        pub           The Windmill
16     26545687        pub        The Queens Head
17     26602504        pub           The Feathers
18     27245287        pub   The Shaftesbury Arms
19     27512718        pub              Lots Road
20     30429954        pub           The Tapestry
21     30429955        pub        Jolly Gardeners
22     30441135        pub               The Hope
23     30478182        pub               The Ship
24     30478187        pub          Ye White Hart
25     30483366        pub        The Park Tavern
26     30483782        pub       Alexandra Tavern
27     30512931        pub           The Cardinal
28     30552863        pub          The Greyhound
29     30629850        pub           The Wych Elm
..          ...        ...                    ...
522  3014221516        pub               The Bear
523  3068306969        pub             The Anchor
524  3133134050        pub           The Red Lion
525  3247663714        pub        The Park Tavern
526  3282498381        pub          The Kings Tun
527  3340684903        pub              The Crown
528  3362469515        pub       The Magic Garden
529  3423539966        pub         The Earlsfield
530  3523526667        pub         The Two Rivers
531  3598403423        pub         The Anchor Pub
532  3681486426        pub       The Old Red Lion
533  3757136233        pub            The Phoenix
534  3770513926        pub      The Jolly Butcher
535  3782616223        pub                 Steins
536  3793532459        pub           The Yardbird
537  3964556370        pub          The Sands End
538  4026677008        pub        Queen Elisabeth
539  4117016441        pub            The Terrace
540  4436916522        pub       Bricklayers Arms
541  4525422451        pub     The Flying Chariot
542  4527208485        pub       Gorgeous Kitchen
543  4527208488        pub         London's Pride
544  4578415857        pub    The Waldegrave Arms
545  4617068622        pub     Ganley's Irish Bar
546  4629007069        pub      The Running Horse
547  4682459836        pub          The Racehorse
548  4926469145        pub  The Merton Apprentice
549  4928682125        pub         The Cricketers
550  5030210975        pub             The Hussar
551  5101640778        pub       The Royal Surrey

[552 rows x 3 columns]

In [13]:
#How many distinct pub names?

df = pd.read_sql_query(
''' 
SELECT COUNT(DISTINCT names.value) as num_pub_names
FROM
    (SELECT id, value
    FROM nodes_tags
    WHERE key='amenity'
    AND value='pub') as pubs 
    JOIN 
    (SELECT id, value
    FROM nodes_tags
    WHERE key='name'
    AND type='regular') as names

ON pubs.id =  names.id


''' , conn)


df

num_pub_names
0            449

In [14]:
# Top 10 Pub Names

df = pd.read_sql_query(
''' 
SELECT names.value, COUNT(names.value) as name_count
FROM
    (SELECT id, value
    FROM nodes_tags
    WHERE key='amenity'
    AND value='pub') as pubs 
    JOIN 
    (SELECT id, value
    FROM nodes_tags
    WHERE key='name'
    AND type='regular') as names

ON pubs.id =  names.id
GROUP BY names.value
ORDER BY name_count DESC
LIMIT 10;

''' , conn)


df

names.value  name_count
0        The Royal Oak           9
1            The Crown           6
2           The Plough           6
3         The Red Lion           5
4           Kings Arms           4
5             The Bell           4
6           The Castle           4
7           The George           4
8  The Prince of Wales           4
9       The White Hart           4

In [15]:

df = pd.read_sql_query(
''' 
SELECT info.value, count(info.value) as count
FROM
    (SELECT id, value
    FROM nodes_tags
    WHERE key='amenity'
    AND value='cafe') as amenities 
    JOIN 
    (SELECT *
    FROM nodes_tags
    WHERE key!='amenity'
    AND type='regular'
    AND key = 'name') as info

ON amenities.id =  info.id
GROUP BY info.value
ORDER BY count DESC
LIMIT 15;

''' , conn)


df




info.value  count
0            Starbucks     30
1                Costa     21
2           Caffè Nero     15
3           Caffe Nero      5
4        Pret A Manger      5
5         Costa Coffee      4
6     Starbucks Coffee      4
7                 Nero      3
8            Puccino's      3
9        Sunshine Cafe      3
10           Best Cafe      2
11           Cafe Nero      2
12  Caffè Nero Express      2
13          Cuppa-cino      2
14      Harris & Hoole      2

In [16]:
df = pd.read_sql_query(
''' 
SELECT info.value, COUNT(info.value) as count
FROM
    (SELECT id, value
    FROM nodes_tags
    WHERE key='amenity'
    AND value='car_sharing') as amenities 
    JOIN 
    (SELECT *
    FROM nodes_tags
    WHERE key!='amenity'
    AND type = 'regular'
    AND key = 'operator') as info

ON amenities.id =  info.id
GROUP BY info.value
ORDER BY count DESC
;

''' , conn)


df

info.value  count
0      Streetcar    140
1         Zipcar     14
2  City Car Club      1
3      StreetCar      1

In [17]:
df = pd.read_sql_query(
''' 
SELECT info.value, COUNT(info.value) as count
FROM
    (SELECT id, value
    FROM nodes_tags
    WHERE key='amenity'
    AND value='fuel') as amenities 
    JOIN 
    (SELECT *
    FROM nodes_tags
    WHERE key!='amenity'
    AND type = 'regular'
    AND key = 'operator') as info

ON amenities.id =  info.id
GROUP BY info.value
ORDER BY count DESC
;

''' , conn)


df

info.value  count
0                  BP     18
1               Shell     15
2                Esso     10
3         Sainsbury's      4
4               Total      3
5               Tesco      2
6                ESSO      1
7                 Jet      1
8    MRH Hatton Cross      1
9               Murco      1
10  Royal Dutch Shell      1
11         Sainsburys      1
12             Texaco      1
13              tesco      1

In [18]:
# Useful query for looking up info available on different amenity types. 
# just change 'value' in first subquery to the amenity you are looking at

df = pd.read_sql_query(
''' 
SELECT *
FROM
    (SELECT id, value
    FROM nodes_tags
    WHERE key='amenity'
    AND value='bicycle_rental') as amenities 
    JOIN 
    (SELECT *
    FROM nodes_tags
    WHERE key!='amenity') as info

ON amenities.id =  info.id
;

''' , conn)


df

id           value          id       key  \
0     261761004  bicycle_rental   261761004  capacity   
1     780114277  bicycle_rental   780114277      name   
2     823375879  bicycle_rental   823375879     local   
3     823375879  bicycle_rental   823375879      name   
4     823375879  bicycle_rental   823375879   network   
5     823375879  bicycle_rental   823375879       ref   
6     836785671  bicycle_rental   836785671  capacity   
7     836785671  bicycle_rental   836785671     local   
8     836785671  bicycle_rental   836785671      name   
9     836785671  bicycle_rental   836785671   network   
10    836785671  bicycle_rental   836785671       ref   
11    836785673  bicycle_rental   836785673  capacity   
12    836785673  bicycle_rental   836785673     local   
13    836785673  bicycle_rental   836785673      name   
14    836785673  bicycle_rental   836785673   network   
15    836785673  bicycle_rental   836785673       ref   
16    836785680  bicycle_rental   836785680  capacity   
17    836785680  bicycle_rental   836785680     local   
18    836785680  bicycle_rental   836785680      name   
19    836785680  bicycle_rental   836785680   network   
20    836785680  bicycle_rental   836785680       ref   
21    836785688  bicycle_rental   836785688  capacity   
22    836785688  bicycle_rental   836785688     local   
23    836785688  bicycle_rental   836785688      name   
24    836785688  bicycle_rental   836785688   network   
25    836785688  bicycle_rental   836785688       ref   
26    836785695  bicycle_rental   836785695  capacity   
27    836785695  bicycle_rental   836785695     local   
28    836785695  bicycle_rental   836785695      name   
29    836785695  bicycle_rental   836785695   network   
..          ...             ...         ...       ...   
145  3581895443  bicycle_rental  3581895443   network   
146  3581895443  bicycle_rental  3581895443  operator   
147  3750681028  bicycle_rental  3750681028      name   
148  3890101248  bicycle_rental  3890101248      name   
149  3890101249  bicycle_rental  3890101249      name   
150  3890126460  bicycle_rental  3890126460      name   
151  3896102441  bicycle_rental  3896102441      name   
152  4112448142  bicycle_rental  4112448142      name   
153  4112448142  bicycle_rental  4112448142   network   
154  4329814892  bicycle_rental  4329814892      name   
155  4537517403  bicycle_rental  4537517403  capacity   
156  4537517403  bicycle_rental  4537517403      name   
157  4537517403  bicycle_rental  4537517403   network   
158  4537517403  bicycle_rental  4537517403  operator   
159  4547084070  bicycle_rental  4547084070      name   
160  4677086112  bicycle_rental  4677086112  capacity   
161  4677086112  bicycle_rental  4677086112      name   
162  4677086112  bicycle_rental  4677086112   network   
163  4677106589  bicycle_rental  4677106589  capacity   
164  4677106589  bicycle_rental  4677106589      name   
165  4677106589  bicycle_rental  4677106589   network   
166  4677109172  bicycle_rental  4677109172  capacity   
167  4677109172  bicycle_rental  4677109172      name   
168  4677109172  bicycle_rental  4677109172   network   
169  4677113221  bicycle_rental  4677113221  capacity   
170  4677113221  bicycle_rental  4677113221      name   
171  4677113221  bicycle_rental  4677113221   network   
172  4677116099  bicycle_rental  4677116099  capacity   
173  4677116099  bicycle_rental  4677116099      name   
174  4677116099  bicycle_rental  4677116099   network   

                             value     type  
0                                2  regular  
1                        Parkcycle  regular  
2                 South Kensington  regular  
3          Gloucester Road Station  regular  
4                   tfl_cycle_hire  regular  
5                          CH19049  regular  
6                               18  regular  
7                     Earl's Court  regular  
8                    Trebovir Road  regular  
9                 

post_box	1607
1	bicycle_parking	792
2	restaurant	663
3	bench	623
4	pub	564
5	cafe	469
6	telephone	438
7	fast_food	321
8	atm	271
9	place_of_worship	271
10	waste_basket	215
11	car_sharing	159
12	pharmacy	155
13	bank	154
14	toilets	150
15	recycling	146
16	parking	132
17	post_office	127
18	fuel	106
19	bar	77
20	bicycle_rental	73
21	dentist	72
22	vending_machine	67
23	school	54
24	doctors	49